File: misc/jobserver_pool.py
#!/usr/bin/env python3
# Copyright 2024 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Setup a GNU Make Jobserver jobs pool then launch a command with it.

On Windows, this only supports the semaphore-based scheme.
On Posix, this uses a pipe by default, unless `--fifo` is used.

On exit, this script verifies that all job slots were returned to
the pool, and will print an error message if this is not the case.

This is useful to catch the use of broken protocol clients.
Use the `--no-check` flag to disable this.

See --help-usage for usage examples.
"""
import argparse
import os
import platform
import subprocess
import sys
import typing as T

# Technical note about the MAKEFLAGS values set by this script.
#
# All the MAKEFLAGS values created by this script begin
# with " -j{count} ", i.e. an initial space, the "-j" characters
# followed by a job slot count then another space.
#
# The initial space is only there to mimic what GNU Make 4.3
# does. Other pool implementations do not use one and thus
# clients should not expect it (even GNU Make doesn't seem
# to care when used as a jobserver client).
#
# The {count} value is also not available in many pool
# implementations, but is useful to better debug multi-builds
# using this script (i.e. to verify that the pool has the
# expected size). Protocol clients should not depend on it
# though.

_DEFAULT_NAME = "jobserver_pool"
_IS_WINDOWS = sys.platform in ("win32", "cygwin")

if _IS_WINDOWS:

    try:
        # This requires pywin32 to be installed.
        import pywintypes
        import win32event
        import win32api
    except ModuleNotFoundError:
        print(
            "\nERROR: Could not import Win32 API, please install pywin32, e.g. `python -m pip install pywin32`.\n",
            file=sys.stderr,
        )
        raise

    # It seems impossible to import a proper mypy-compatible type definition for PyHANDLE
    # 'from pywintypes import PyHANDLE' fails stating there is no such name.
    # 'from pywintypes import HANDLE as PyHANDLE' fails because HANDLE is a function, not a type.
    PyHandle: T.TypeAlias = T.Any

    def create_sem(
        sem_name: str, jobs_count: int
    ) -> T.Tuple[PyHandle, T.Dict[str, str]]:
        """Create and initialize Win32 semaphore."""
        assert jobs_count > 0, f"Jobs count must be strictly positive"
        # The win32event documentation states that the first argument to CreateSemaphore()
        # can be None to indicate default security attributes, but mypy only wants
        # a PySECURITY_ATTRIBUTES for some reason.
        handle = win32event.CreateSemaphore(
            None,  # type: ignore
            jobs_count - 1,
            jobs_count - 1,
            sem_name,
        )
        assert bool(handle), f"Error creating Win32 semaphore {win32api.GetLastError()}"
        # See technical note above about MAKEFLAGS format.
        env = dict(os.environ)
        env["MAKEFLAGS"] = f" -j{jobs_count} --jobserver-auth=" + sem_name
        return handle, env

    def check_sem_count(handle: PyHandle, jobs_count: int) -> int:
        if jobs_count <= 1:
            # Nothing to check here.
            return 0

        expected_count = jobs_count - 1

        read_count = win32event.ReleaseSemaphore(handle, 1)
        if read_count < expected_count:
            print(
                f"ERROR: {expected_count - read_count} were missing from the jobs pool (got {read_count}, expected {expected_count})",
                file=sys.stderr,
            )
            return 1
        if read_count > expected_count:
            print(
                f"ERROR: {read_count - expected_count} extra tokens were released to the jobs pool (got {read_count}, expected {expected_count})",
                file=sys.stderr,
            )
            return 1

        return 0

    def print_usage() -> int:
        print(
            r"""Example usage:

# Start <command> after setting the server to provide as many jobs
# as available CPUs (the default)
python \path\to\jobserver_pool.py <command>

# Start <command> with a fixed number of job slots.
python \path\to\jobserver_pool.py -j10 <command>

# Disable the feature with a non-positive count. This is equivalent
# to running <command> directly.
python \path\to\jobserver_pool.py -j0 <command>

# Use a specific semaphore name
python \path\to\jobserver_pool.py --name=my_build_jobs <command>

# Setup jobserver then start new interactive PowerShell
# session, print MAKEFLAGS value, build stuff, then exit.
python \path\to\jobserver_pool.py powershell.exe
$env:MAKEFLAGS
... build stuff ...
exit
"""
        )
        return 0

else:  # !_IS_WINDOWS

    def create_pipe(jobs_count: int) -> T.Tuple[int, int, T.Dict[str, str]]:
        """Create and fill Posix PIPE."""
        read_fd, write_fd = os.pipe()
        os.set_inheritable(read_fd, True)
        os.set_inheritable(write_fd, True)
        assert jobs_count > 0, f"Token count must be strictly positive"
        os.write(write_fd, (jobs_count - 1) * b"x")
        # See technical note above about MAKEFLAGS format.
        env = dict(os.environ)
        env["MAKEFLAGS"] = (
            f" -j{jobs_count} --jobserver-fds={read_fd},{write_fd} --jobserver-auth={read_fd},{write_fd}"
        )
        return read_fd, write_fd, env

    def create_fifo(path: str, jobs_count: int) -> T.Tuple[int, int, T.Dict[str, str]]:
        """Create and fill Posix FIFO."""
        if os.path.exists(path):
            os.remove(path)

        # mypy complains that this does not exit on Windows.
        os.mkfifo(path)  # type: ignore

        read_fd = os.open(path, os.O_RDONLY | os.O_NONBLOCK)
        write_fd = os.open(path, os.O_WRONLY | os.O_NONBLOCK)
        assert jobs_count > 0, f"Token count must be strictly positive"
        os.write(write_fd, (jobs_count - 1) * b"x")
        # See technical note above about MAKEFLAGS format.
        env = dict(os.environ)
        env["MAKEFLAGS"] = f" -j{jobs_count} --jobserver-auth=fifo:" + path
        return read_fd, write_fd, env

    def print_usage() -> int:
        print(
            r"""Example usage:

# Start <command> after setting the job pool to provide as many jobs
# as available CPUs (the default)
/path/to/jobserver_pool.py <command>

# Start <command> with a fixed number of jobs
/path/to/jobserver_pool.py -j10 <command>

# Disable the feature with a non-positive count. This is equivalent
# to running <command> directly.
/path/to/jobserver_pool.py -j0 <command>

# Use a specific FIFO path
/path/to/jobserver_pool.py --fifo=/tmp/my_build_jobs <command>

# Setup jobserver then start new interactive Bash shell
# session, print MAKEFLAGS value, build stuff, then exit.
/path/to/jobserver_pool.py bash -i
echo "$MAKEFLAGS"
... build stuff ...
exit
"""
        )
        return 0

    def check_pipe_tokens(read_fd: int, jobs_count: int) -> int:
        if jobs_count <= 1:  # Nothing to check
            return 0

        # Remove implicit token from the expected count.
        expected_count = jobs_count - 1
        os.set_blocking(read_fd, False)
        read_count = 0
        while True:
            try:
                token = os.read(read_fd, 1)
                if len(token) == 0:  # End of pipe?
                    break
                read_count += 1
            except BlockingIOError:
                break

        if read_count < expected_count:
            print(
                f"ERROR: {expected_count - read_count} tokens were missing from the jobs pool (got {read_count}, expected {expected_count})",
                file=sys.stderr,
            )
            return 1
        if read_count > expected_count:
            print(
                f"ERROR: {read_count - expected_count} extra tokens were released to the jobs pool (got {read_count}, expected {expected_count})",
                file=sys.stderr,
            )
            return 1

        return 0


def main() -> int:
    parser = argparse.ArgumentParser(
        description=__doc__, formatter_class=argparse.RawTextHelpFormatter
    )
    if _IS_WINDOWS:
        parser.add_argument(
            "--name",
            help=f"Specify semaphore name, default is {_DEFAULT_NAME}",
            default=_DEFAULT_NAME,
        )
    else:
        mutex_group = parser.add_mutually_exclusive_group()
        mutex_group.add_argument(
            "--pipe",
            action="store_true",
            help="Implement the pool with a Unix pipe (the default)",
        )
        mutex_group.add_argument(
            "--fifo",
            help=f"Implement the pool with a FIFO file path. E,g, ./{_DEFAULT_NAME}",
        )

    parser.add_argument(
        "--no-check",
        action="store_true",
        help="Disable the final check that verifies that all job slots were returned to the pool on exit.",
    )

    parser.add_argument(
        "--help-usage", action="store_true", help="Print usage examples."
    )

    parser.add_argument(
        "-j",
        "--jobs",
        action="store",
        metavar="COUNT",
        dest="jobs_count",
        type=int,
        default=os.cpu_count(),
        help="Set job slots ccount, default is available CPUs count",
    )

    parser.add_argument("command", nargs=argparse.REMAINDER, help="Command to run.")
    args = parser.parse_args()

    if args.help_usage:
        return print_usage()

    if not args.command:
        parser.error("This script requires at least one command argument!")

    jobs_count = args.jobs_count
    if jobs_count <= 0:
        # Disable the feature.
        ret = subprocess.run(args.command)
        exit_code = ret.returncode
    elif _IS_WINDOWS:
        # Run with a Window semaphore.
        try:
            handle, env = create_sem(args.name, jobs_count)
            ret = subprocess.run(args.command, env=env)
            exit_code = ret.returncode

            if exit_code == 0 and not args.no_check:
                exit_code = check_sem_count(handle, jobs_count)

        finally:
            win32api.CloseHandle(handle)
    else:
        # Run with pipe descriptors by default, or a FIFO if --fifo is used.
        exit_code = 0
        fifo_path = ""
        try:
            if args.fifo:
                fifo_path = os.path.abspath(args.fifo)
                read_fd, write_fd, env = create_fifo(fifo_path, args.jobs_count)
                ret = subprocess.run(args.command, env=env)
            else:
                read_fd, write_fd, env = create_pipe(args.jobs_count)
                ret = subprocess.run(
                    args.command, env=env, pass_fds=(read_fd, write_fd)
                )

            exit_code = ret.returncode
            if exit_code == 0 and not args.no_check:
                exit_code = check_pipe_tokens(read_fd, jobs_count)

        finally:
            os.close(read_fd)
            os.close(write_fd)

            if fifo_path:
                os.remove(fifo_path)

    return exit_code


if __name__ == "__main__":
    sys.exit(main())

File: src/jobserver_pool.cc
// Copyright 2025 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "jobserver_pool.h"

#include <assert.h>

#include "util.h"

#ifdef _WIN32

#include <windows.h>

class Win32JobserverPool : public JobserverPool {
 public:
  static std::unique_ptr<Win32JobserverPool> Create(size_t slot_count,
                                                    std::string* error) {
    assert(slot_count > 1 && "slot_count must be 2 or higher");
    auto pool = std::unique_ptr<Win32JobserverPool>(new Win32JobserverPool());
    if (!pool->InitWithSemaphore(slot_count, error))
      pool.reset();
    return pool;
  }

  std::string GetEnvMakeFlagsValue() const override {
    std::string result;
    result.resize(sem_name_.size() + 32);
    int ret =
        snprintf(const_cast<char*>(result.data()), result.size(),
                 " -j%zd --jobserver-auth=%s", job_count_, sem_name_.c_str());
    if (ret < 0 || ret > static_cast<int>(result.size()))
      Fatal("Could not format Win32JobserverPool MAKEFLAGS!");

    return result;
  }

  virtual ~Win32JobserverPool() {
    if (IsValid())
      ::CloseHandle(handle_);
  }

 private:
  Win32JobserverPool() = default;

  // CreateSemaphore returns NULL on failure.
  bool IsValid() const {
    // CreateSemaphoreA() returns NULL on failure, not INVALID_HANDLE_VALUE.
    return handle_ != NULL;
  }

  // Compute semaphore name for new instance.
  static std::string GetSemaphoreName() {
    // Use a per-process global counter to allow multiple instances of this
    // class to run in the same process. Useful for unit-tests.
    static int counter = 0;
    counter += 1;
    char name[64];
    snprintf(name, sizeof(name), "ninja_jobserver_pool_%d_%d",
             GetCurrentProcessId(), counter);
    return std::string(name);
  }

  bool InitWithSemaphore(size_t slot_count, std::string* error) {
    job_count_ = slot_count;
    sem_name_ = GetSemaphoreName();
    LONG count = static_cast<LONG>(slot_count - 1);
    handle_ = ::CreateSemaphoreA(NULL, count, count, sem_name_.c_str());
    if (!IsValid()) {
      *error = "Could not create semaphore: " + GetLastErrorString();
      return false;
    }
    return true;
  }

  // Semaphore handle.
  HANDLE handle_ = NULL;

  // Saved slot count.
  size_t job_count_ = 0;

  // Semaphore name.
  std::string sem_name_;
};

#else  // !_WIN32

#include <errno.h>
#include <fcntl.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

class PosixJobserverPool : public JobserverPool {
 public:
  static std::unique_ptr<PosixJobserverPool> Create(size_t slot_count,
                                                    std::string* error) {
    assert(slot_count > 1 && "slot_count must be 2 or higher");
    auto pool = std::unique_ptr<PosixJobserverPool>(new PosixJobserverPool());
    if (!pool->InitWithFifo(slot_count, error)) {
      pool.reset();
    }
    return pool;
  }

  std::string GetEnvMakeFlagsValue() const override {
    std::string result;
    if (!fifo_.empty()) {
      result.resize(fifo_.size() + 32);
      int ret = snprintf(const_cast<char*>(result.data()), result.size(),
                         " -j%zd --jobserver-auth=fifo:%s", job_count_,
                         fifo_.c_str());
      if (ret < 0 || ret > static_cast<int>(result.size()))
        Fatal("Could not format PosixJobserverPool MAKEFLAGS!");
      result.resize(static_cast<size_t>(ret));
    }
    return result;
  }

  virtual ~PosixJobserverPool() {
    if (write_fd_ >= 0)
      ::close(write_fd_);
    if (!fifo_.empty())
      ::unlink(fifo_.c_str());
  }

 private:
  PosixJobserverPool() = default;

  // Fill the pool to satisfy |slot_count| job slots. This
  // writes |slot_count - 1| bytes to the pipe to satisfy the
  // implicit job slot requirement.
  bool FillSlots(size_t slot_count, std::string* error) {
    job_count_ = slot_count;
    for (; slot_count > 1; --slot_count) {
      // Write '+' into the pipe, just like GNU Make. Note that some
      // implementations write '|' instead, but so far no client or pool
      // implementation cares about the exact value, though the official spec
      // says this might change in the future.
      const char slot_char = '+';
      ssize_t ret = ::write(write_fd_, &slot_char, 1);
      if (ret != 1) {
        if (ret < 0 && errno == EINTR)
          continue;
        *error =
            std::string("Could not fill job slots pool: ") + strerror(errno);
        return false;
      }
    }
    return true;
  }

  bool InitWithFifo(size_t slot_count, std::string* error) {
    const char* tmp_dir = getenv("TMPDIR");
    if (!tmp_dir)
      tmp_dir = "/tmp";

    fifo_.resize(strlen(tmp_dir) + 32);
    int len = snprintf(const_cast<char*>(fifo_.data()), fifo_.size(),
                       "%s/NinjaFIFO%d", tmp_dir, getpid());
    if (len < 0) {
      *error = "Cannot create fifo path!";
      return false;
    }
    fifo_.resize(static_cast<size_t>(len));

    int ret = mknod(fifo_.c_str(), S_IFIFO | 0666, 0);
    if (ret < 0) {
      *error = std::string("Cannot create fifo: ") + strerror(errno);
      return false;
    }

    do {
      write_fd_ = ::open(fifo_.c_str(), O_RDWR | O_CLOEXEC);
    } while (write_fd_ < 0 && errno == EINTR);
    if (write_fd_ < 0) {
      *error = std::string("Could not open fifo: ") + strerror(errno);
      // Let destructor remove the fifo.
      return false;
    }
    return FillSlots(slot_count, error);
  }

  // Number of parallel job slots (including implicit one).
  size_t job_count_ = 0;

  // A non-inheritable file descriptor to keep the pool alive.
  int write_fd_ = -1;

  // Path to fifo, this will be empty when using an anonymous pipe.
  std::string fifo_;
};
#endif  // !_WIN32

// static
std::unique_ptr<JobserverPool> JobserverPool::Create(size_t num_job_slots,
                                                     std::string* error) {
  if (num_job_slots < 2) {
    *error = "At least 2 job slots needed";
    return nullptr;
  }

#ifdef _WIN32
  return Win32JobserverPool::Create(num_job_slots, error);
#else   // !_WIN32
  return PosixJobserverPool::Create(num_job_slots, error);
#endif  // !_WIN32
}
